front-end/src/components/Chart.js

```diff
@@ -249,14 +263,104 @@ const Chart = ({ chartInformation }) => {
+            setIsLoaded(false)
+            console.log("window unloaded")
+        }
+    }, [chartInformation])
+
+    useEffect(() => {
+        if (!chartInformation.window || !isLoaded) return;
+    
+        const videoElement = chartInformation.window.document.getElementById('video');
+        console.log("videoElement:", videoElement)
+        if (!videoElement) return;
+
+        const handleTimeUpdate = () => {
+            setVideoTimestamp(videoElement.currentTime * 1000);
+        };
+        
+        videoElement.addEventListener('timeupdate', handleTimeUpdate);
+        
+        return () => {
+            videoElement.removeEventListener('timeupdate', handleTimeUpdate);
+        };
+    }, [isLoaded]);
+
+    const [offsets, setOffsets] = useState([])
+
+    useEffect(() => {
+        if (chartInformation.video.fileHeaders.length == 0) return
+        const videoStart = new Date(chartInformation.video.fileHeaders[0]).getTime()
+        
+        const tempOffsets = []
+        chartInformation.files.forEach(file => {
+            const fileStart = new Date(file.columns[0].timespan[0]).getTime() // Unix date of first timestamp in file
+            tempOffsets.push(videoStart - fileStart)
+        })
+
+        setOffsets(tempOffsets)
+    }, [chartInformation])
+
+    useEffect(() => {
+        if (chartRef.current.chart.series.length == 0) return
+
+        const pointIndexs = []
+        chartRef.current.chart.series.forEach(series => {
+            const seriesIndex = chartRef.current.chart.series.indexOf(series)
+
+            const fileTimestamp = videoTimestamp + offsets[seriesIndex] + timestamps[0]
+            if (fileTimestamp < timestamps[0] || fileTimestamp > timestamps[timestamps.length - 1]) return
+            
+            const timestampIndex = findClosestTimestamp(fileTimestamp)
+            const pointIndex = findPointIndex(timestampIndex, series)
+            if (pointIndex >= 0) pointIndexs.push({series: seriesIndex, point: pointIndex})
+        })
+
+        if (pointIndexs.length === 0) return
+        chartRef.current.chart.series[pointIndexs[0].series].points[pointIndexs[0].point].onMouseOver()
+        if (pointIndexs.length > 1) pointIndexs.slice(1).forEach(pointIndex => {
+            chartRef.current.chart.series[pointIndex.series].points[pointIndex.point].setState('hover')
+        })
+        
+    }, [videoTimestamp])
+
+
+    const findClosestTimestamp = (timestamp) => {
+
+        const closestTimestamp = timestamps.reduce((prev, curr) => Math.abs(curr - timestamp) < Math.abs(prev - timestamp) ? curr : prev)
+
+        return timestamps.indexOf(closestTimestamp);
+    }
+
+    const findPointIndex = (timestampIndex, series) => {
+        const timestampPoint = {x: series.xData[timestampIndex], y: series.yData[timestampIndex]}
+        const point = series.points.find(point => point.x === timestampPoint.x && point.y === timestampPoint.y)
+        return series.points.indexOf(point)
+    }
+
```

Comment:
Can we separate all the functions above into a separate file and store them in a lib folder? See either Jack's branch or multiple graphs, and make a new file for that folder so merge conflicts dont arise

Reply:
yea we should put these in a utils function

Reply:
lib folder was created